### PR TITLE
Abstracting the creation of netty channel handlers

### PR DESCRIPTION
**Pull request by nsivabalan**
Creation of handlers was hardcoded in NettyServer. This patch abstracts it in the NettyServerFactory so that any other implementation of NioServer have the flexibility to define their own handlers.

SLA: 5 mins
Reviewers: Priyesh, Ming

### CODE REVIEW

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 1]
```diff
@@ -49,27 +50,20 @@
   private final NettyMetrics nettyMetrics;
   private final EventLoopGroup bossGroup;
   private final EventLoopGroup workerGroup;
-  private final RestRequestHandler requestHandler;
-  private final PublicAccessLogger publicAccessLogger;
-  private final RestServerState restServerState;
+  private final ArrayList<Map.Entry<String, ChannelHandler>> channelHandlers;
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: This looks like a hack. I assume the intention here is a list of pairs? You shouldn't have to use `Map.Entry` for that. There are no keys and values and there are no maps in this context.
I suggest creating a small class of your own with meaningful names for the tuple/pair.

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: +1

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: The class should be `public` so that it can be accessed by other implementations of the `NettyServerFactory`.

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: I'd be ok with a `Map` too but looks like @pnarayanan might prefer the class. I would say there is a key value relationship with the name being the key and the value being the handler.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 2]
```diff
@@ -49,27 +50,20 @@
...
 
   /**
    * Creates a new instance of NettyServer.
    * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
    * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
-   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
-   * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
-   * @param restServerState the {@link RestServerState} that can be used to check the health of the system
-   *                              to respond to health check requests
+   * @param channelHandlers list of {@link io.netty.channel.ChannelHandler}s to be used in the channel pipeline
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: Don't need fully qualified name for `ChannelHandler`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 3]
```diff
@@ -49,27 +50,20 @@
...
    */
-  public NettyServer(NettyConfig nettyConfig, NettyMetrics nettyMetrics, RestRequestHandler requestHandler,
-      PublicAccessLogger publicAccessLogger, RestServerState restServerState) {
+  public NettyServer(NettyConfig nettyConfig, NettyMetrics nettyMetrics,
+      ArrayList<Map.Entry<String, ChannelHandler>> channelHandlers) {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: `ArrayList` is too restrictive. Please make it a `List`

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java (hunk 1 above)
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: Though in the current usage there is no strict key-value relationship, is there? There is an association between these entries, sure, but it is not strictly key-value as nothing stops you from having multiple keys with the same value in this `ArrayList`. That said, would a `Map` itself work here? Why not use that?

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: You bring up a good point - we shouldn't have anything with the same keys which means a `Map` is ideal here.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 4]
```diff
@@ -49,27 +49,21 @@
   private final NettyMetrics nettyMetrics;
   private final EventLoopGroup bossGroup;
   private final EventLoopGroup workerGroup;
-  private final RestRequestHandler requestHandler;
-  private final PublicAccessLogger publicAccessLogger;
-  private final RestServerState restServerState;
+  private final LinkedHashMap<String, ChannelHandler> channelHandlerInfoList;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: not a list anymore?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 5]
```diff
@@ -49,27 +49,21 @@
...
 
   /**
    * Creates a new instance of NettyServer.
    * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
    * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
-   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
-   * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
-   * @param restServerState the {@link RestServerState} that can be used to check the health of the system
-   *                              to respond to health check requests
+   * @param channelHandlerInfoList Linkedin list of pairs of {@link ChannelHandler} name and {@link ChannelHandler}s to
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: Linkedin? :)

**Comment by vgkholla** [comment]
Minor comments. LGTM. Please make sure you have built and formatted.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java [hunk 6]
```diff
@@ -30,9 +35,8 @@
 
   private final NettyConfig nettyConfig;
   private final NettyMetrics nettyMetrics;
-  private final RestRequestHandler requestHandler;
-  private final PublicAccessLogger publicAccessLogger;
-  private final RestServerState restServerState;
+  // linked hashmap as we need a deterministic order while iterating through the map
+  private LinkedHashMap<String, ChannelHandler> channelHandlerInfoList;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: `final`. You can do the `new LinkedHashMap()` right here.

**Comment by vgkholla** [comment]
Just a single comment. LGTM

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java [hunk 7]
```diff
@@ -78,11 +80,39 @@ public NettyServerFactory(VerifiableProperties verifiableProperties, MetricRegis
   }
 
   /**
+   * Initialize the {@link ChannelHandler}s to be used in the netty pipeline
+   * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: reorder the params.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java [hunk 8]
```diff
@@ -116,6 +121,39 @@ private NettyServer getNettyServer(Properties properties)
...
     PublicAccessLogger publicAccessLogger = new PublicAccessLogger(new String[]{}, new String[]{});
     RestServerState restServerState = new RestServerState("/healthCheck");
-    return new NettyServer(nettyConfig, nettyMetrics, requestHandler, publicAccessLogger, restServerState);
+    return new NettyServer(nettyConfig, nettyMetrics,
+        initializeChannelHandlers(nettyMetrics, nettyConfig, requestHandler, publicAccessLogger, restServerState));
+  }
+
+  /**
+   * Initialize the {@link ChannelHandler}s to be used in the netty pipeline
+   * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
```
Comment (pnarayanan) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java]: reorder.

**Comment by pnarayanan** [comment]
Looks good. Just a couple of minor comments.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java [hunk 9]
```diff
@@ -78,11 +80,39 @@ public NettyServerFactory(VerifiableProperties verifiableProperties, MetricRegis
...
+   * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
+   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
+   * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
+   * @param restServerState the {@link RestServerState} that can be used to check the health of the system
+   *                              to respond to health check requests
+   */
+  private void initializeChannelHandlers(NettyMetrics nettyMetrics, NettyConfig nettyConfig,
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: The purpose of this patch is to introduce flexibility when creating pipeline. It does so in NettyServer constructor, however, you will still need to have this method hard coded. Possible that we just supply a list of handles to this initializer without fixing the parameters?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java [hunk 10]
```diff
@@ -116,6 +121,39 @@ private NettyServer getNettyServer(Properties properties)
...
+   * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
+   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
```
Comment (xiahome) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java]: the {@link RestRequestHandler} that handles general requests.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java [hunk 11]
```diff
@@ -78,11 +80,39 @@ public NettyServerFactory(VerifiableProperties verifiableProperties, MetricRegis
   }
 
   /**
+   * Initialize the {@link ChannelHandler}s to be used in the netty pipeline
+   * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
+   * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
+   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: the {@link RestRequestHandler} that handles general requests.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 12]
```diff
@@ -49,27 +49,21 @@
...
 
   /**
    * Creates a new instance of NettyServer.
    * @param nettyConfig the {@link NettyConfig} instance that defines the configuration parameters for the NettyServer.
    * @param nettyMetrics the {@link NettyMetrics} instance to use to record metrics.
-   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
-   * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
-   * @param restServerState the {@link RestServerState} that can be used to check the health of the system
-   *                              to respond to health check requests
+   * @param channelHandlerInfoList Linked list of pairs of {@link ChannelHandler} name and {@link ChannelHandler}s to
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: An ordered list of pairs of {@link ChannelHandler} name and {@link ChannelHandler}s...

**Comment by xiahome** [comment]
A high-level question and some minor comments. The rest LGTM.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java (hunk 9 above)
Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: Good point. We are providing two levels of flexibility. One at the Factory Level and one at the Server Level. Either users can define their own NioServer and NioServerFactory. Or they can use the readily available NettyServer with their own custom handlers. So, this patch is about the latter.

**Comment by nsivabalan** [comment]
@pnarayanan @vgkholla @xiahome : Can you guys take a look at it

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java [hunk 13]
```diff
@@ -36,14 +36,14 @@
...
-  private LinkedHashMap<String, ChannelHandler> channelHandlerInfoList;
+  private final LinkedHashMap<String, ChannelHandler> channelHandlerInfoList = new LinkedHashMap<>();
   private final Logger logger = LoggerFactory.getLogger(getClass());
 
   /**
    * Creates a new instance of NettyServerFactory.
    * @param verifiableProperties the in-memory {@link VerifiableProperties} to use.
    * @param metricRegistry the {@link MetricRegistry} to use.
-   * @param requestHandler the {@link RestRequestHandler} that can be used to submit requests that need to be handled.
+   * @param requestHandler the {@link RestRequestHandler} that handles general requests.
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServerFactory.java]: not sure why this change was made. What is "general" requests?

**Comment by vgkholla** [comment]
LGTM

**Comment by xiahome** [comment]
LGTM, merging.
